cpp/src/strings/char_types/char_cases.cu

```diff
@@ -0,0 +1,268 @@
+      739,    743,    751,    757,    761,    769,    773,    787,    797,    809,
+      811,    821,    823,    827,    829,    839,    853,    857,    859,    863,
+      877,    881,    883,    887,    907,    911,    919,    929,    937,    941,
+      947,    953,    967,    971,    977,    983,    991,    997,   1009,   1013,
+    1019,   1021,   1031,   1033,   1039,   1049,   1051,   1061,   1063,   1069,
+    1087,   1091,   1093,   1097,   1103,   1109,   1117,   1123,   1129,   1151,
+    1153,   1163,   1171,   1181,   1187,   1193,   1201,   1213,   1217,   1223,
+    1229,   1231,   1237,   1249,   1259,   1277,   1279,   1283,   1289,   1291,
+    1297,   1301,   1303,   1307,   1319,   1321,   1327,   1361,   1367,   1373,
+    1381,   1399,   1409,   1423,   1427,   1429,   1433,   1439,   1447,   1451,
+    1453,   1459,   1471,   1481,   1483,   1487,   1489,   1493,   1499,   1511,
+    1523,   1531,   1543,   1549,   1553,   1559,   1567,   1571,   1579,   1583,
+    1597,   1601,   1607,   1609,   1613,   1619,   1621,   1627,   1637,   1657,
+    1663,   1667,   1669,   1693,   1697,   1699,   1709,   1721,   1723,   1733,
+    1741,   1747,   1753,   1759,   1777,   1783,   1787,   1789,   1801,   1811,
+    1823,   1831,   1847,   1861,   1867,   1871,   1873,   1877,   1879,   1889,
+    1901,   1907,   1913,   1931,   1933,   1949,   1951,   1973,   1979,   1987,
+    1993,   1997,   1999,   2003,   2011,   2017,   2027,   2029,   2039,   2053,
+    2063,   2069,   2081,   2083,   2087,   2089,   2099,
+  };
+
+  // find a prime number that generates no collisions for all possible input data
+  uint16_t find_collision_proof_prime(std::vector<uint16_t> const& data)
+  {
+    for( auto pitr=primes.begin(); pitr < primes.end(); ++pitr ){
+        std::map<uint16_t,uint16_t> keys;
+        for( auto itr=data.begin(); itr<data.end(); ++itr){
+          int key = *itr % *pitr;
+          if( keys.find(key) == keys.end() ){
+              keys[key] = *itr;
+          }
+        }
+        if( keys.size() == data.size() ){
+          return *pitr;
+        }            
+    }
+
+    // couldn't find a collision-proof prime
+    return 0;
+  }
+
+} // anonymous namespace
+
+/**
+ * @copydoc cudf::strings::detail::generate_special_mapping_hash_table 
+ */
+void generate_special_mapping_hash_table()
+{
+  uint16_t hash_prime = find_collision_proof_prime(codepoints_in);
+  if(hash_prime == 0){
+      CUDF_FAIL("Could not find a usable prime number for hash table");
+  }
+
+  // generate hash index table   
+  // size of the table is the prime #, since we're just doing (key % hash_prime)
+  std::vector<uint16_t> hash_indices(hash_prime);      
+  int count = 0;
+  std::for_each(codepoints_in.begin(), codepoints_in.end(), [&](uint16_t codepoint){      
+      hash_indices[codepoint % hash_prime] = count;
+      count++;
+  });
+
+  // print out the code
+
+  // the mappings
+  printf("struct special_case_mapping {\n");
```

Comment:
Is this for debug?

Reply:
Sort of.  It's for regenerating the hash table of character mappings.  You wouldn't be calling this during normal cudf operation, but if you had to update the mappings in cudf itself (eg for a newer version of Unicode) , you would end up running this, taking the output and pasting it back into char_cases.h

Reply:
If that's really important, why not make it a program that is run by cmake to generate the header before building? (like we do for stringify)

Reply:
This already sort of happens.  The input to this code (`special_case_mapping_in`, `codepoints_in`) came from some external Spark code.    The code you see here is in fact run ahead of time and generates  `g_special_case_mappings`, `g_special_case_hash_indices` and `get_special_case_hash_index`

The overall thinking is as follows:
- Leaving the hash table generation code in cudf itself makes it easier for someone who may have to regenerate a new hash table later (eg, for a new version of Unicode) rather than having to recreate it.

- By leaving the input generate seperate, it would allow different mechanisms for generating the list of mappings (not everyone has easy access to Spark).